src/vs/workbench/browser/parts/editor/multiEditorTabsControl.ts

```diff
@@ -889,29 +890,30 @@ export class MultiEditorTabsControl extends EditorTabsControl {
 			if (editor) {
 				if (e.shiftKey) {
 					let anchor;
-					if (this.lastSelectedEditor && this.groupView.isSelected(this.lastSelectedEditor)) {
+					if (this.lastSingleSelectSelectedEditor && this.tabsModel.isSelected(this.lastSingleSelectSelectedEditor)) {
 						// The last selected editor is the anchor
-						anchor = this.lastSelectedEditor;
+						anchor = this.lastSingleSelectSelectedEditor;
 					} else {
 						// The active editor is the anchor
-						this.lastSelectedEditor = this.groupView.activeEditor!;
+						this.lastSingleSelectSelectedEditor = this.groupView.activeEditor!;
 						anchor = this.groupView.activeEditor!;
 					}
-					this.selectEditorsBetween(editor, anchor);
+					await this.selectEditorsBetween(editor, anchor);
 				} else if ((e.ctrlKey && !isMacintosh) || (e.metaKey && isMacintosh)) {
-					if (this.groupView.isSelected(editor)) {
-						this.groupView.unSelectEditor(editor);
+					if (this.tabsModel.isSelected(editor)) {
+						await this.unselectEditor(editor);
 					} else {
-						this.groupView.selectEditor(editor, true);
-						this.lastSelectedEditor = editor;
+						await this.selectEditor(editor);
+						this.lastSingleSelectSelectedEditor = editor;
 					}
 				} else {
 					// Even if focus is preserved make sure to activate the group.
-					this.groupView.openEditor(editor, { preserveFocus, activation: EditorActivation.ACTIVATE }, { selected: this.groupView.isSelected(editor) /* Ensures drag and drop does not remove selection */ });
+					// If new active editor is selected, keep the current selection on key down
+					// such that drag and drop can operate over the selection. Selection is remove on key up in this case.
+					const selection = this.groupView.isSelected(editor) ? [this.groupView.activeEditor!, ...this.groupView.selectedEditors.filter(e => !e.matches(editor))] : [];
```

Comment:
Sometimes you use `groupView.isSelected` and sometimes `tabsModel.isSelected`. I think if possible every call should go to the model, no? Maybe check throughtout the file.

Reply:
Should always be `tabsModel.isSelected`